File: apps/api/src/modules/reviews/application/reviews.service.ts
import { Inject, Injectable } from '@nestjs/common';
import type { ReviewRecommendationFilter, ReviewSort } from '@gamescore/shared';
import type {
  CreateReviewRequest,
  CursorPaginatedResponse,
  ReviewDto,
  ReviewVoteResponse,
  UpdateReviewRequest,
} from '@gamescore/types';
import type { ReportReason } from '@gamescore/shared';

import { AppConfigService } from '../../../common/config/app-config.service';
import { hashClientIp, isoDate, reviewFingerprint } from '../../../common/crypto/hash';
import {
  BadRequestError,
  ConflictError,
  ForbiddenError,
  NotFoundError,
} from '../../../common/errors/app.exception';
import { ERROR_CODES } from '../../../common/errors/error-codes';
import { clampLimit, cursorMeta, DEFAULT_CURSOR_SIZE } from '../../../common/http/pagination';
import { JOB_NAMES } from '../../../common/jobs/job-names';
import { JOB_QUEUE, type JobQueue } from '../../../common/jobs/job-queue';
import { PrismaService } from '../../../common/prisma/prisma.service';
import type { AuthUser } from '../../auth/domain/auth-user';
import { GameRepository } from '../../games/repositories/game.repository';
import { ReputationService } from '../../ratings/application/reputation.service';
import { ReviewRankingService } from '../../ratings/application/review-ranking.service';
import { ScoreRecalculationService } from '../../ratings/application/score-recalculation.service';
import { toReviewDto } from '../mappers/review.mapper';
import { ReviewRepository, type ReviewWithRelations } from '../repositories/review.repository';
import { AntiSpamService } from './anti-spam.service';

@Injectable()
export class ReviewsService {
  constructor(
    private readonly reviews: ReviewRepository,
    private readonly games: GameRepository,
    private readonly antiSpam: AntiSpamService,
    private readonly ranking: ReviewRankingService,
    private readonly reputation: ReputationService,
    private readonly scores: ScoreRecalculationService,
    private readonly prisma: PrismaService,
    private readonly config: AppConfigService,
    @Inject(JOB_QUEUE) private readonly jobs: JobQueue,
  ) {}

  async listForGame(
    slug: string,
    query: {
      sort?: ReviewSort;
      recommendation?: ReviewRecommendationFilter;
      platformId?: string;
      minHoursPlayed?: number;
      maxHoursPlayed?: number;
      cursor?: string;
      limit?: number;
    },
    viewer?: AuthUser | null,
  ): Promise<CursorPaginatedResponse<ReviewDto>> {
    const game = await this.requireGame(slug);
    const limit = clampLimit(query.limit, DEFAULT_CURSOR_SIZE);
    const { items, nextCursor } = await this.reviews.list({
      gameId: game.id,
      sort: query.sort ?? 'BEST',
      recommendation: query.recommendation ?? 'ALL',
      platformId: query.platformId,
      minHoursPlayed: query.minHoursPlayed,
      maxHoursPlayed: query.maxHoursPlayed,
      viewerId: viewer?.id,
      cursor: query.cursor,
      limit,
    });

    return {
      items: await this.withViewerContext(items, viewer?.id),
      meta: cursorMeta(limit, nextCursor),
    };
  }

  async getById(id: string, viewer?: AuthUser | null): Promise<ReviewDto> {
    const review = await this.requireReview(id);
    const [mapped] = await this.withViewerContext([review], viewer?.id);
    return mapped!;
  }

  async create(
    slug: string,
    input: CreateReviewRequest,
    author: AuthUser,
    clientIp?: string,
  ): Promise<ReviewDto> {
    const game = await this.requireGame(slug);
    const existing = await this.reviews.findActiveByUserAndGame(author.id, game.id);
    if (existing) {
      throw new ConflictError(
        ERROR_CODES.REVIEW_ALREADY_EXISTS,
        'You already have a review for this game',
      );
    }

    await this.assertPlatform(game.id, input.platformId);
    this.assertReviewPayload(input);

    const fingerprint = reviewFingerprint(input.text);
    const ipHash = hashClientIp(clientIp, this.config.jwt.accessSecret);
    const spam = await this.antiSpam.inspect({
      userId: author.id,
      text: input.text,
      fingerprint,
      ipHash,
    });

    const rankingScore = this.ranking.score({
      usefulVotes: 0,
      notUsefulVotes: 0,
      authorReputation: author.reputationScore,
      textLength: input.text.trim().length,
      hoursPlayed: input.hoursPlayed,
      rating: input.rating,
      createdAt: new Date(),
    });

    const created = await this.prisma.$transaction(async (tx) => {
      const review = await this.reviews.create(
        {
          gameId: game.id,
          userId: author.id,
          recommended: input.recommended,
          text: input.text.trim(),
          rating: input.rating ?? null,
          hoursPlayed: input.hoursPlayed ?? null,
          platformId: input.platformId ?? null,
          rankingScore,
          textFingerprint: fingerprint,
          authorIpHash: ipHash,
          moderationStatus: spam.moderationStatus,
        },
        tx,
      );

      await this.reputation.apply(
        author.id,
        'REVIEW_PUBLISHED',
        { type: 'review', id: review.id },
        tx,
      );
      await this.scores.bumpDailyActivity(game.id, input.recommended, review.createdAt, 1, tx);
      await this.scores.recalculate(game.id, tx);
      return review;
    });

    await this.jobs.enqueue(JOB_NAMES.DETECT_REVIEW_BOMB, {
      gameId: game.id,
      date: isoDate(created.createdAt),
    });
    await this.jobs.enqueue(JOB_NAMES.INVALIDATE_RANKINGS, {});

    const [mapped] = await this.withViewerContext([created], author.id);
    return mapped!;
  }

  async update(id: string, input: UpdateReviewRequest, actor: AuthUser): Promise<ReviewDto> {
    const review = await this.requireReview(id);
    if (review.userId !== actor.id) {
      throw new ForbiddenError(ERROR_CODES.REVIEW_NOT_OWNED, 'You can only edit your own review');
    }
    if (review.deletedAt) {
      throw new NotFoundError(ERROR_CODES.REVIEW_NOT_FOUND, 'Review not found');
    }

    this.assertReviewPayload({
      recommended: input.recommended ?? review.recommended,
      text: input.text ?? review.text,
      rating: input.rating === undefined ? review.rating : input.rating,
      hoursPlayed: input.hoursPlayed === undefined ? review.hoursPlayed : input.hoursPlayed,
    });

    if (input.platformId !== undefined) {
      await this.assertPlatform(review.gameId, input.platformId);
    }

    const nextText = input.text?.trim() ?? review.text;
    const fingerprint = reviewFingerprint(nextText);
    const rankingScore = this.ranking.score({
      usefulVotes: review.usefulCount,
      notUsefulVotes: review.notUsefulCount,
      authorReputation: actor.reputationScore,
      textLength: nextText.length,
      hoursPlayed: input.hoursPlayed === undefined ? review.hoursPlayed : input.hoursPlayed,
      rating: input.rating === undefined ? review.rating : input.rating,
      createdAt: review.createdAt,
    });

    const previousRecommended = review.recommended;
    const nextRecommended = input.recommended ?? review.recommended;

    const updated = await this.prisma.$transaction(async (tx) => {
      const saved = await this.reviews.update(
        id,
        {
          recommended: nextRecommended,
          text: nextText,
          rating: input.rating === undefined ? undefined : input.rating,
          hoursPlayed: input.hoursPlayed === undefined ? undefined : input.hoursPlayed,
          platformId: input.platformId === undefined ? undefined : input.platformId,
          edited: true,
          textFingerprint: fingerprint,
          rankingScore,
        },
        tx,
      );

      if (previousRecommended !== nextRecommended) {
        await this.scores.bumpDailyActivity(review.gameId, previousRecommended, review.createdAt, -1, tx);
        await this.scores.bumpDailyActivity(review.gameId, nextRecommended, review.createdAt, 1, tx);
      }

      await this.scores.recalculate(review.gameId, tx);
      return saved;
    });

    await this.jobs.enqueue(JOB_NAMES.INVALIDATE_RANKINGS, {});
    const [mapped] = await this.withViewerContext([updated], actor.id);
    return mapped!;
  }

  async softDelete(id: string, actor: AuthUser, reason?: string): Promise<void> {
    const review = await this.requireReview(id);
    const isOwner = review.userId === actor.id;
    const isStaff = actor.role === 'MODERATOR' || actor.role === 'ADMIN';
    if (!isOwner && !isStaff) {
      throw new ForbiddenError(ERROR_CODES.REVIEW_NOT_OWNED, 'You can only delete your own review');
    }
    if (review.deletedAt) {
      throw new NotFoundError(ERROR_CODES.REVIEW_NOT_FOUND, 'Review not found');
    }

    await this.prisma.$transaction(async (tx) => {
      await this.reviews.update(
        id,
        {
          deletedAt: new Date(),
          deletedById: actor.id,
          deletionReason: reason ?? (isOwner ? 'Deleted by author' : 'Removed by moderator'),
          status: 'HIDDEN',
        },
        tx,
      );

      await this.reputation.apply(
        review.userId,
        isOwner ? 'REVIEW_DELETED' : 'REVIEW_REMOVED_BY_MODERATOR',
        { type: 'review', id: review.id },
        tx,
      );
      await this.scores.bumpDailyActivity(review.gameId, review.recommended, review.createdAt, -1, tx);
      await this.scores.recalculate(review.gameId, tx);
    });

    await this.jobs.enqueue(JOB_NAMES.INVALIDATE_RANKINGS, {});
  }

  async vote(id: string, useful: boolean, voter: AuthUser): Promise<ReviewVoteResponse> {
    const review = await this.requireReview(id);
    if (review.userId === voter.id) {
      throw new ForbiddenError(ERROR_CODES.CANNOT_VOTE_OWN_REVIEW, 'You cannot vote on your own review');
    }
    if (review.deletedAt || review.status !== 'PUBLISHED') {
      throw new NotFoundError(ERROR_CODES.REVIEW_NOT_FOUND, 'Review not found');
    }

    const result = await this.prisma.$transaction(async (tx) => {
      const existing = await this.reviews.findVote(id, voter.id, tx);

      if (existing?.useful === useful) {
        return this.reviews.findById(id, tx);
      }

      if (existing) {
        await this.reviews.upsertVote(id, voter.id, useful, tx);
        await this.reputation.apply(
          review.userId,
          existing.useful ? 'USEFUL_VOTE_REMOVED' : 'NOT_USEFUL_VOTE_REMOVED',
          { type: 'review', id },
          tx,
        );
        await this.reputation.apply(
          review.userId,
          useful ? 'USEFUL_VOTE_RECEIVED' : 'NOT_USEFUL_VOTE_RECEIVED',
          { type: 'review', id },
          tx,
        );
        await this.reviews.update(
          id,
          {
            usefulCount: { increment: useful ? 1 : -1 },
            notUsefulCount: { increment: useful ? -1 : 1 },
          },
          tx,
        );
      } else {
        await this.reviews.upsertVote(id, voter.id, useful, tx);
        await this.reputation.apply(
          review.userId,
          useful ? 'USEFUL_VOTE_RECEIVED' : 'NOT_USEFUL_VOTE_RECEIVED',
          { type: 'review', id },
          tx,
        );
        await this.reviews.update(
          id,
          useful ? { usefulCount: { increment: 1 } } : { notUsefulCount: { increment: 1 } },
          tx,
        );
      }

      const updated = await this.reviews.findById(id, tx);
      if (updated) {
        const rankingScore = this.ranking.score({
          usefulVotes: updated.usefulCount,
          notUsefulVotes: updated.notUsefulCount,
          authorReputation: updated.user.reputationScore,
          textLength: updated.text.length,
          hoursPlayed: updated.hoursPlayed,
          rating: updated.rating,
          createdAt: updated.createdAt,
        });
        await this.ranking.persist(id, rankingScore, tx);
        return this.reviews.findById(id, tx);
      }
      return updated;
    });

    if (!result) {
      throw new NotFoundError(ERROR_CODES.REVIEW_NOT_FOUND, 'Review not found');
    }

    return {
      reviewId: result.id,
      usefulCount: result.usefulCount,
      notUsefulCount: result.notUsefulCount,
      viewerVote: useful ? 'USEFUL' : 'NOT_USEFUL',
    };
  }

  async removeVote(id: string, voter: AuthUser): Promise<ReviewVoteResponse> {
    const review = await this.requireReview(id);
    const existing = await this.reviews.findVote(id, voter.id);
    if (!existing) {
      throw new NotFoundError(ERROR_CODES.VOTE_NOT_FOUND, 'You have not voted on this review');
    }

    const result = await this.prisma.$transaction(async (tx) => {
      await this.reviews.deleteVote(id, voter.id, tx);
      await this.reputation.apply(
        review.userId,
        existing.useful ? 'USEFUL_VOTE_REMOVED' : 'NOT_USEFUL_VOTE_REMOVED',
        { type: 'review', id },
        tx,
      );
      await this.reviews.update(
        id,
        existing.useful
          ? { usefulCount: { decrement: 1 } }
          : { notUsefulCount: { decrement: 1 } },
        tx,
      );
      return this.reviews.findById(id, tx);
    });

    return {
      reviewId: id,
      usefulCount: result?.usefulCount ?? 0,
      notUsefulCount: result?.notUsefulCount ?? 0,
      viewerVote: null,
    };
  }

  async report(
    id: string,
    input: { reason: ReportReason; details?: string },
    reporter: AuthUser,
  ): Promise<void> {
    const review = await this.requireReview(id);
    if (review.userId === reporter.id) {
      throw new ForbiddenError(
        ERROR_CODES.CANNOT_REPORT_OWN_REVIEW,
        'You cannot report your own review',
      );
    }

    const existing = await this.reviews.findReport(id, reporter.id);
    if (existing) {
      throw new ConflictError(ERROR_CODES.REVIEW_ALREADY_REPORTED, 'You have already reported this review');
    }

    await this.reviews.createReport({
      reviewId: id,
      reporterId: reporter.id,
      reason: input.reason,
      details: input.details,
    });
  }

  private async withViewerContext(
    items: ReviewWithRelations[],
    viewerId?: string,
  ): Promise<ReviewDto[]> {
    const context = viewerId
      ? await this.reviews.viewerContext(
          items.map((item) => item.id),
          viewerId,
        )
      : { votes: new Map<string, boolean>(), reports: new Set<string>() };

    return items.map((item) =>
      toReviewDto(item, {
        viewerId,
        viewerVote: context.votes.has(item.id) ? context.votes.get(item.id)! : null,
        viewerHasReported: context.reports.has(item.id),
      }),
    );
  }

  private async requireGame(slug: string) {
    const game = await this.games.findBySlug(slug);
    if (!game) {
      throw new NotFoundError(ERROR_CODES.GAME_NOT_FOUND, 'Game not found');
    }
    return game;
  }

  private async requireReview(id: string) {
    const review = await this.reviews.findById(id);
    if (!review) {
      throw new NotFoundError(ERROR_CODES.REVIEW_NOT_FOUND, 'Review not found');
    }
    return review;
  }

  private async assertPlatform(gameId: string, platformId?: string | null): Promise<void> {
    if (!platformId) return;
    const game = await this.games.findById(gameId);
    const allowed = game?.platforms.some((link) => link.platformId === platformId);
    if (!allowed) {
      throw new BadRequestError(
        ERROR_CODES.PLATFORM_NOT_AVAILABLE_FOR_GAME,
        'That platform is not listed for this game',
      );
    }
  }

  private assertReviewPayload(input: {
    recommended: boolean;
    text: string;
    rating?: number | null;
    hoursPlayed?: number | null;
  }): void {
    if (input.rating != null && (input.rating < 0 || input.rating > 10)) {
      throw new BadRequestError(ERROR_CODES.VALIDATION_FAILED, 'Rating must be between 0 and 10');
    }
    if (input.hoursPlayed != null && (input.hoursPlayed < 0 || input.hoursPlayed > 100_000)) {
      throw new BadRequestError(ERROR_CODES.VALIDATION_FAILED, 'Hours played is out of range');
    }
    if (!input.text || input.text.trim().length === 0) {
      throw new BadRequestError(ERROR_CODES.VALIDATION_FAILED, 'Review text is required');
    }
  }
}

File: apps/api/src/GameScore.Infrastructure/Reviews/ReviewsService.cs
using GameScore.Application.Configuration;
using GameScore.Application.Errors;
using GameScore.Application.Http;
using GameScore.Infrastructure.Auth;
using GameScore.Infrastructure.Crypto;
using GameScore.Infrastructure.Integrations;
using GameScore.Infrastructure.Jobs;
using GameScore.Infrastructure.Mappers;
using GameScore.Infrastructure.Persistence;
using GameScore.Infrastructure.Ratings;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Options;

namespace GameScore.Infrastructure.Reviews;

public sealed class ReviewsService
{
    private readonly GameScoreDbContext _db;
    private readonly ReviewDataAccess _reviews;
    private readonly AntiSpamService _antiSpam;
    private readonly ReviewRankingService _ranking;
    private readonly ReputationService _reputation;
    private readonly ScoreRecalculationService _scores;
    private readonly GameImportService _importer;
    private readonly BackgroundJobScheduler _jobs;
    private readonly AppConfig _config;

    public ReviewsService(
        GameScoreDbContext db,
        ReviewDataAccess reviews,
        AntiSpamService antiSpam,
        ReviewRankingService ranking,
        ReputationService reputation,
        ScoreRecalculationService scores,
        GameImportService importer,
        BackgroundJobScheduler jobs,
        IOptions<AppConfig> config)
    {
        _db = db;
        _reviews = reviews;
        _antiSpam = antiSpam;
        _ranking = ranking;
        _reputation = reputation;
        _scores = scores;
        _importer = importer;
        _jobs = jobs;
        _config = config.Value;
    }

    public async Task<object> ListForGameSlugAsync(
        string slug,
        string sort,
        string recommendation,
        Guid? platformId,
        int? minHours,
        int? maxHours,
        string? cursor,
        int? limit,
        Guid? viewerId,
        CancellationToken cancellationToken = default)
    {
        var game = await _db.Games.AsNoTracking().FirstOrDefaultAsync(g => g.Slug == slug, cancellationToken);
        if (game is null)
        {
            throw new NotFoundException(ErrorCodes.GameNotFound, "Game not found");
        }

        var pageSize = Pagination.ClampLimit(limit, Pagination.DefaultCursorSize);
        var (items, nextCursor) = await _reviews.ListForGameAsync(
            game.Id,
            sort,
            recommendation,
            platformId,
            minHours,
            maxHours,
            cursor,
            pageSize,
            cancellationToken);

        return new
        {
            items = await MapManyAsync(items, viewerId, cancellationToken),
            meta = Pagination.CursorMeta(pageSize, nextCursor),
        };
    }

    public async Task<object> GetByIdAsync(Guid id, Guid? viewerId, CancellationToken cancellationToken = default)
    {
        var bundle = await _reviews.LoadAsync(id, cancellationToken);
        if (bundle is null || bundle.Review.DeletedAt is not null)
        {
            throw new NotFoundException(ErrorCodes.ReviewNotFound, "Review not found");
        }

        return (await MapManyAsync([bundle], viewerId, cancellationToken))[0];
    }

    public Task<object> CreateAsync(
        string slug,
        Guid userId,
        int reputationScore,
        bool recommended,
        int? rating,
        string text,
        int? hoursPlayed,
        Guid? platformId,
        string? clientIp,
        CancellationToken cancellationToken = default) =>
        CreateInternalAsync(slug, null, userId, reputationScore, recommended, rating, text, hoursPlayed, platformId, clientIp, cancellationToken);

    public async Task<object> CreateForExternalAsync(
        string externalId,
        Guid userId,
        int reputationScore,
        bool recommended,
        int? rating,
        string text,
        int? hoursPlayed,
        Guid? platformId,
        string? clientIp,
        CancellationToken cancellationToken = default)
    {
        var imported = await _importer.ImportByExternalIdAsync(externalId, cancellationToken);
        var resolvedPlatform = await ResolvePlatformRefAsync(imported.GameId, platformId, cancellationToken);
        return await CreateInternalAsync(
            imported.Slug,
            imported.GameId,
            userId,
            reputationScore,
            recommended,
            rating,
            text,
            hoursPlayed,
            resolvedPlatform,
            clientIp,
            cancellationToken);
    }

    public async Task<object> UpdateAsync(
        Guid id,
        Guid userId,
        int reputationScore,
        bool? recommended,
        int? rating,
        string? text,
        int? hoursPlayed,
        Guid? platformId,
        CancellationToken cancellationToken = default)
    {
        var bundle = await _reviews.LoadTrackedAsync(id, cancellationToken);
        if (bundle is null || bundle.Review.DeletedAt is not null)
        {
            throw new NotFoundException(ErrorCodes.ReviewNotFound, "Review not found");
        }

        if (bundle.Review.UserId != userId)
        {
            throw new ForbiddenException(ErrorCodes.ReviewNotOwned, "You can only edit your own reviews");
        }

        var review = bundle.Review;
        var nextRecommended = recommended ?? review.Recommended;
        var nextText = text?.Trim() ?? review.Text;
        AssertPayload(nextRecommended, nextText, rating ?? review.Rating, hoursPlayed ?? review.HoursPlayed);

        if (platformId is not null)
        {
            await ResolvePlatformRefAsync(review.GameId, platformId, cancellationToken);
            review.PlatformId = platformId;
        }

        var fingerprint = HashUtil.ReviewFingerprint(nextText);
        var moderation = await _antiSpam.InspectAsync(
            new AntiSpamInput(userId, nextText, fingerprint, review.AuthorIpHash, SkipRateLimits: true, ExcludeReviewId: review.Id),
            cancellationToken);

        var rankingScore = _ranking.Score(
            review.UsefulCount,
            review.NotUsefulCount,
            reputationScore,
            nextText.Length,
            hoursPlayed ?? review.HoursPlayed,
            rating ?? review.Rating,
            review.CreatedAt);

        var previousRecommended = review.Recommended;
        review.Recommended = nextRecommended;
        review.Text = nextText;
        if (rating is not null)
        {
            review.Rating = rating;
        }

        if (hoursPlayed is not null)
        {
            review.HoursPlayed = hoursPlayed;
        }

        review.Edited = true;
        review.TextFingerprint = fingerprint;
        review.ModerationStatus = moderation;
        review.RankingScore = rankingScore;
        review.UpdatedAt = DateTime.UtcNow;

        await using var tx = await _db.Database.BeginTransactionAsync(cancellationToken);
        if (previousRecommended != nextRecommended)
        {
            await _scores.BumpDailyActivityAsync(review.GameId, previousRecommended, review.CreatedAt, -1, cancellationToken);
            await _scores.BumpDailyActivityAsync(review.GameId, nextRecommended, review.CreatedAt, 1, cancellationToken);
        }

        await _scores.RecalculateInsideAsync(review.GameId, includePlatforms: true, cancellationToken);
        await _db.SaveChangesAsync(cancellationToken);
        await tx.CommitAsync(cancellationToken);

        _jobs.EnqueueRecalculateGameScore(review.GameId);
        _jobs.EnqueueInvalidateRankings();
        await _reputation.InvalidateUserStatsAsync(userId, cancellationToken);

        return (await MapManyAsync([await _reviews.LoadAsync(id, cancellationToken)!], userId, cancellationToken))[0];
    }

    public async Task SoftDeleteAsync(Guid id, Guid userId, string? reason, CancellationToken cancellationToken = default)
    {
        var bundle = await _reviews.LoadTrackedAsync(id, cancellationToken);
        if (bundle is null || bundle.Review.DeletedAt is not null)
        {
            throw new NotFoundException(ErrorCodes.ReviewNotFound, "Review not found");
        }

        if (bundle.Review.UserId != userId)
        {
            throw new ForbiddenException(ErrorCodes.ReviewNotOwned, "You can only delete your own reviews");
        }

        var review = bundle.Review;
        review.DeletedAt = DateTime.UtcNow;
        review.DeletedById = userId;
        review.DeletionReason = reason ?? "Deleted by author";
        review.Status = DbReviewStatus.HIDDEN;
        review.UpdatedAt = DateTime.UtcNow;

        var author = await _db.Users.FirstAsync(u => u.Id == review.UserId, cancellationToken);
        await using var tx = await _db.Database.BeginTransactionAsync(cancellationToken);
        _reputation.ApplyToUser(author, "REVIEW_DELETED", "review", review.Id);
        await _scores.BumpDailyActivityAsync(review.GameId, review.Recommended, review.CreatedAt, -1, cancellationToken);
        await _scores.RecalculateInsideAsync(review.GameId, true, cancellationToken);
        await _db.SaveChangesAsync(cancellationToken);
        await tx.CommitAsync(cancellationToken);

        _jobs.EnqueueRecalculateGameScore(review.GameId);
        _jobs.EnqueueInvalidateRankings();
        await _reputation.InvalidateUserStatsAsync(review.UserId, cancellationToken);
    }

    public async Task<object> VoteAsync(Guid id, bool useful, Guid voterId, CancellationToken cancellationToken = default)
    {
        var bundle = await _reviews.LoadTrackedAsync(id, cancellationToken);
        if (bundle is null || bundle.Review.DeletedAt is not null || bundle.Review.Status != DbReviewStatus.PUBLISHED)
        {
            throw new NotFoundException(ErrorCodes.ReviewNotFound, "Review not found");
        }

        if (bundle.Review.UserId == voterId)
        {
            throw new ForbiddenException(ErrorCodes.CannotVoteOwnReview, "You cannot vote on your own review");
        }

        var review = bundle.Review;
        var author = await _db.Users.FirstAsync(u => u.Id == review.UserId, cancellationToken);
        var existing = await _db.ReviewVotes.FirstOrDefaultAsync(v => v.ReviewId == id && v.UserId == voterId, cancellationToken);

        await using var tx = await _db.Database.BeginTransactionAsync(cancellationToken);
        if (existing?.Useful == useful)
        {
            await _db.SaveChangesAsync(cancellationToken);
            await tx.CommitAsync(cancellationToken);
            return VoteResponse(review, useful);
        }

        if (existing is not null)
        {
            existing.Useful = useful;
            existing.UpdatedAt = DateTime.UtcNow;
            _reputation.ApplyToUser(author, existing.Useful ? "USEFUL_VOTE_REMOVED" : "NOT_USEFUL_VOTE_REMOVED", "review", id);
            _reputation.ApplyToUser(author, useful ? "USEFUL_VOTE_RECEIVED" : "NOT_USEFUL_VOTE_RECEIVED", "review", id);
            review.UsefulCount = Math.Max(0, review.UsefulCount + (useful ? 1 : -1));
            review.NotUsefulCount = Math.Max(0, review.NotUsefulCount + (useful ? -1 : 1));
        }
        else
        {
            _db.ReviewVotes.Add(new ReviewVoteEntity
            {
                Id = Guid.NewGuid(),
                ReviewId = id,
                UserId = voterId,
                Useful = useful,
                CreatedAt = DateTime.UtcNow,
                UpdatedAt = DateTime.UtcNow,
            });
            _reputation.ApplyToUser(author, useful ? "USEFUL_VOTE_RECEIVED" : "NOT_USEFUL_VOTE_RECEIVED", "review", id);
            if (useful)
            {
                review.UsefulCount++;
            }
            else
            {
                review.NotUsefulCount++;
            }
        }

        review.RankingScore = _ranking.Score(
            review.UsefulCount,
            review.NotUsefulCount,
            author.ReputationScore,
            review.Text.Length,
            review.HoursPlayed,
            review.Rating,
            review.CreatedAt);
        review.UpdatedAt = DateTime.UtcNow;
        await _db.SaveChangesAsync(cancellationToken);
        await tx.CommitAsync(cancellationToken);

        await _reputation.InvalidateUserStatsAsync(review.UserId, cancellationToken);
        return VoteResponse(review, useful);
    }

    public async Task<object> RemoveVoteAsync(Guid id, Guid voterId, CancellationToken cancellationToken = default)
    {
        var existing = await _db.ReviewVotes.FirstOrDefaultAsync(v => v.ReviewId == id && v.UserId == voterId, cancellationToken);
        if (existing is null)
        {
            throw new NotFoundException(ErrorCodes.VoteNotFound, "You have not voted on this review");
        }

        var review = await _db.Reviews.FirstAsync(r => r.Id == id, cancellationToken);
        var author = await _db.Users.FirstAsync(u => u.Id == review.UserId, cancellationToken);

        await using var tx = await _db.Database.BeginTransactionAsync(cancellationToken);
        _db.ReviewVotes.Remove(existing);
        _reputation.ApplyToUser(author, existing.Useful ? "USEFUL_VOTE_REMOVED" : "NOT_USEFUL_VOTE_REMOVED", "review", id);
        if (existing.Useful)
        {
            review.UsefulCount = Math.Max(0, review.UsefulCount - 1);
        }
        else
        {
            review.NotUsefulCount = Math.Max(0, review.NotUsefulCount - 1);
        }

        review.UpdatedAt = DateTime.UtcNow;
        await _db.SaveChangesAsync(cancellationToken);
        await tx.CommitAsync(cancellationToken);
        await _reputation.InvalidateUserStatsAsync(review.UserId, cancellationToken);

        return new
        {
            reviewId = id,
            usefulCount = review.UsefulCount,
            notUsefulCount = review.NotUsefulCount,
            viewerVote = (string?)null,
        };
    }

    public async Task ReportAsync(
        Guid id,
        Guid reporterId,
        string reason,
        string? details,
        CancellationToken cancellationToken = default)
    {
        var review = await _db.Reviews.AsNoTracking().FirstOrDefaultAsync(r => r.Id == id, cancellationToken);
        if (review is null)
        {
            throw new NotFoundException(ErrorCodes.ReviewNotFound, "Review not found");
        }

        if (review.UserId == reporterId)
        {
            throw new ForbiddenException(ErrorCodes.CannotReportOwnReview, "You cannot report your own review");
        }

        if (await _db.ReviewReports.AnyAsync(r => r.ReviewId == id && r.ReporterId == reporterId, cancellationToken))
        {
            throw new ConflictException(ErrorCodes.ReviewAlreadyReported, "You have already reported this review");
        }

        _db.ReviewReports.Add(new ReviewReportEntity
        {
            Id = Guid.NewGuid(),
            ReviewId = id,
            ReporterId = reporterId,
            Reason = Enum.Parse<DbReportReason>(reason),
            Details = details,
            Status = DbReportStatus.PENDING,
            CreatedAt = DateTime.UtcNow,
        });
        await _db.SaveChangesAsync(cancellationToken);
    }

    private async Task<object> CreateInternalAsync(
        string slug,
        Guid? gameId,
        Guid userId,
        int reputationScore,
        bool recommended,
        int? rating,
        string text,
        int? hoursPlayed,
        Guid? platformId,
        string? clientIp,
        CancellationToken cancellationToken)
    {
        gameId ??= (await _db.Games.AsNoTracking().FirstOrDefaultAsync(g => g.Slug == slug, cancellationToken))?.Id;
        if (gameId is null)
        {
            throw new NotFoundException(ErrorCodes.GameNotFound, "Game not found");
        }

        if (await _db.Reviews.AnyAsync(r => r.UserId == userId && r.GameId == gameId && r.DeletedAt == null, cancellationToken))
        {
            throw new ConflictException(ErrorCodes.ReviewAlreadyExists, "You already have a review for this game");
        }

        platformId = await ResolvePlatformRefAsync(gameId.Value, platformId, cancellationToken);
        AssertPayload(recommended, text, rating, hoursPlayed);

        var fingerprint = HashUtil.ReviewFingerprint(text);
        var ipHash = HashUtil.HashClientIp(clientIp, _config.Jwt.AccessSecret);
        var moderation = await _antiSpam.InspectAsync(
            new AntiSpamInput(userId, text, fingerprint, ipHash),
            cancellationToken);

        var now = DateTime.UtcNow;
        var rankingScore = _ranking.Score(0, 0, reputationScore, text.Trim().Length, hoursPlayed, rating, now);
        var review = new ReviewEntity
        {
            Id = Guid.NewGuid(),
            GameId = gameId.Value,
            UserId = userId,
            Recommended = recommended,
            Rating = rating,
            Text = text.Trim(),
            HoursPlayed = hoursPlayed,
            PlatformId = platformId,
            RankingScore = rankingScore,
            ModerationStatus = moderation,
            TextFingerprint = fingerprint,
            AuthorIpHash = ipHash,
            CreatedAt = now,
            UpdatedAt = now,
        };

        _db.Reviews.Add(review);
        var author = await _db.Users.FirstAsync(u => u.Id == userId, cancellationToken);

        await using var tx = await _db.Database.BeginTransactionAsync(cancellationToken);
        _reputation.ApplyToUser(author, "REVIEW_PUBLISHED", "review", review.Id);
        await _scores.BumpDailyActivityAsync(gameId.Value, recommended, now, 1, cancellationToken);
        await _scores.RecalculateInsideAsync(gameId.Value, true, cancellationToken);
        await _db.SaveChangesAsync(cancellationToken);
        await tx.CommitAsync(cancellationToken);

        _jobs.EnqueueDetectReviewBomb(gameId.Value, DateHelpers.IsoDate(now));
        _jobs.EnqueueRecalculateGameScore(gameId.Value);
        _jobs.EnqueueInvalidateRankings();
        await _reputation.InvalidateUserStatsAsync(userId, cancellationToken);

        var bundle = await _reviews.LoadAsync(review.Id, cancellationToken);
        return (await MapManyAsync([bundle!], userId, cancellationToken))[0];
    }

    private static object VoteResponse(ReviewEntity review, bool useful) =>
        new
        {
            reviewId = review.Id,
            usefulCount = review.UsefulCount,
            notUsefulCount = review.NotUsefulCount,
            viewerVote = useful ? "USEFUL" : "NOT_USEFUL",
        };

    private async Task<IReadOnlyList<object>> MapManyAsync(
        IReadOnlyList<ReviewBundle> items,
        Guid? viewerId,
        CancellationToken cancellationToken)
    {
        var ids = items.Select(i => i.Review.Id).ToList();
        var context = viewerId is null
            ? (Votes: new Dictionary<Guid, bool>(), Reports: new HashSet<Guid>())
            : await _reviews.ViewerContextAsync(ids, viewerId.Value, cancellationToken);

        return items.Select(item =>
        {
            bool? viewerVote = null;
            if (viewerId is not null && context.Votes.TryGetValue(item.Review.Id, out var vote))
            {
                viewerVote = vote;
            }

            return ReviewMapper.ToDto(
                item.Review,
                item.Author,
                item.Game,
                item.Platform,
                viewerId,
                viewerVote,
                context.Reports.Contains(item.Review.Id));
        }).ToList();
    }

    private async Task<Guid?> ResolvePlatformRefAsync(Guid gameId, Guid? platformRef, CancellationToken cancellationToken)
    {
        if (platformRef is null)
        {
            return null;
        }

        var link = await _db.GamePlatforms.AsNoTracking()
            .Include(gp => gp.Platform)
            .FirstOrDefaultAsync(
                gp => gp.GameId == gameId && (gp.PlatformId == platformRef || gp.Platform.Slug == platformRef.ToString()),
                cancellationToken);

        if (link is null)
        {
            var bySlug = await _db.GamePlatforms.AsNoTracking()
                .Include(gp => gp.Platform)
                .FirstOrDefaultAsync(gp => gp.GameId == gameId && gp.Platform.Slug == platformRef.ToString(), cancellationToken);
            link = bySlug;
        }

        if (link is null)
        {
            throw new BadRequestException(
                ErrorCodes.PlatformNotAvailableForGame,
                "That platform is not listed for this game");
        }

        return link.PlatformId;
    }

    private static void AssertPayload(bool recommended, string text, int? rating, int? hoursPlayed)
    {
        if (rating is < 0 or > 10)
        {
            throw new BadRequestException(ErrorCodes.ValidationFailed, "Rating must be between 0 and 10");
        }

        if (hoursPlayed is < 0 or > 100_000)
        {
            throw new BadRequestException(ErrorCodes.ValidationFailed, "Hours played is out of range");
        }

        if (string.IsNullOrWhiteSpace(text))
        {
            throw new BadRequestException(ErrorCodes.ValidationFailed, "Review text is required");
        }
    }
}
